source/common/buffer/buffer_impl.h

```diff
@@ -1,16 +1,98 @@
 #pragma once
 
 #include <cstdint>
+#include <deque>
 #include <string>
 
 #include "envoy/buffer/buffer.h"
 
 #include "common/common/non_copyable.h"
-#include "common/event/libevent.h"
 
 namespace Envoy {
 namespace Buffer {
 
+/**
+ * A BufferSlice manages a contiguous block of bytes.
+ * The block is arranged like this:
+ *                   |<- data_size() -->|                    |<- reservable_size() ->|
+ * +-----------------+------------------+--------------------+-----------------------+
+ * | Drained         | Data             | Reserved           | Reservable            |
+ * | Unused space    | Usable content   | Additional content | Unused space that can |
+ * | that formerly   |                  | can be added here  | can be moved into     |
+ * | was in the Data |                  |                    | the Reserved section  |
+ * | section         |                  |                    |                       |
+ * +-----------------+------------------+--------------------+-----------------------+
+ *                   ^
+ *                   |
+ *                   data()
+ */
+class BufferSlice {
+public:
+  using Reservation = std::pair<void*, uint64_t>;
```

Comment:
Instead of std::pair, use absl::string_view? It is more readable than `.first`/`.second` and comes with some utility function.

Reply:
Ah I think you're right string_view is const pointer. Then perhaps a simple struct with non-const pointer and size will work, I think in general we prefer named struct than std::pair.

Reply:
+1 on struct vs pair. Thus I suggest
```
struct Reservation {
  uint8_t* data_;
  uint64_t size_;
};
```
This may increase readability by reducing casts? See also RawSlice in include/envoy/buffer/buffer.h which is similar, but uses void* so will require casting for all uses.

Reply:
Maybe RawSlice itself should be the data type that represents a reservation. I'm thinking through the pros and cons now...